protocol/deployment-chains.md

```diff
@@ -0,0 +1,261 @@
+an output can be JSON, and each account entry overwrites the contents of the previous account entry.
+For simplicity we avoid storage-merging, and just overwrite accounts as a whole.
+We do not have contracts with particularly many storage entries, nor do the medium-sized ones change with every step. 
+
+By performing merges this way, we can also make hardforks clean:
+e.g. we can produce a state for fork A, and then fork it and apply the upgrade to fork B, producing a tiny diff.
+
+The smaller the files, the more we can afford, and the higher the limit
+for configuration test-coverage is.
+
+##### How do we not miss the cache with every redeployment?
+
+It is important to omit the "genesis timestamp" from the cache-key, and use some fixed value as input instead:
+every real-time test will need a different timestamp, so embedding it as input would break all caching.
+Luckily, only the legacy L2-output-oracle L1 contract enshrines anything,
+and the L2 genesis block is easily modified before chain-specs are generated.
+
+##### How does this work with production deployments?
+
+The caching, chaining, and commiting is not a part of production.
+As it is against a live chain, not a local state-dump.
+
+However, the isolated configuration and deployment-sub-routines do enable
+a deployer to plug into the superchain contracts, while customizing just what they need,
+without having to untangle a monolithic single deployment script.
+
+Ideally this is performed through the OPStackManager.
+
+### Result
+
+This approach enables us to get rid of "mega-feature" env-vars (Plasma, fault-proofs, Interop)
+as well as allocs hardfork-suffixes: everything is configuration,
+and we can just fork of existing configuration to tweak something.
+
+The main changes to the deploy/genesis-scripting are to:
+1. Get rid of L1 and L2 "allocs": every incremental output is a small "allocs", identified by the hash of inputs.
+2. Get rid of env-vars: we make all customization the same type of well-defined input.
+3. Split the deploy-config: each deployment sub-routine should have its own well-contained config,
+  such that we can attribute the output to the change of input.
+
+The Go `op-e2e` side should then be able to produce the needed configuration input-groups, hash them,
+and then either hit the cache of outputs, or invoke the solidity-scripts to generate the outputs. 
+
+Backwards compatibility can be achieved by adding a function that loads a legacy deploy-config,
+and applies the sequence of all deployment steps, to generate what would have previously been generated.
+
+The OPStackManager spec and draft already strongly point the L2-instance-specific
+deployment of L1 contracts to a subset of deploy-configuration.
+At this time that subset is under-specified (misses configuration, no fault-proof support, etc.),
+but it could utilize the L2-instance specific input-group and deployment routine, to achieve what it needs.
```

Comment:
Yea this is correct, the spec is currently focused on deploying standard chains specifically. It removes all unnecessary inputs and only exposes allowable standard config params towards that goal.

An open question is how to expand it to support other configurations, with the rough idea mentioned above in https://github.com/ethereum-optimism/design-docs/pull/52/files?diff=unified&w=0#r1692407267. Architecting the hooks or flags based on the resulting input groups that come out of this is likely a good solution